ltm/inserted_context_agent.py

```diff
@@ -0,0 +1,577 @@
+        #     if "trivia" in kw:
+        #         trivia_skip = True
+        #
+        # if trivia_skip:
+        #     llm_filtered_mems = keyword_filtered_mems
+        # else:
+        #     llm_filtered_mems = self.llm_memory_filter(keyword_filtered_mems, query_dict["queries"], all_keywords, cost_cb)
+
+        llm_filtered_mems = self.llm_memory_filter(keyword_filtered_mems, query_dict["queries"], all_keywords, cost_cb)
+
+        sorted_mems = sorted(llm_filtered_mems, key=lambda x: x.timestamp)
+        return sorted_mems
+
+    def memory_present(self, memory, memory_list):
+        # passage_info seems to be unique to memory, regardless of the query
+        for list_mem in memory_list:
+            if memory.passage_info.fromIndex == list_mem.passage_info.fromIndex and memory.passage_info.toIndex == list_mem.passage_info.toIndex:
+                return True
+        return False
+
+    def save_interaction(self, user_message, response_message, timestamp, keywords):
+        self.semantic_memory.add_text(user_message, timestamp=timestamp, metadata={"keywords": keywords})
+        self.semantic_memory.add_separator()
+        self.semantic_memory.add_text(response_message, timestamp=timestamp, metadata={"keywords": keywords})
+        self.semantic_memory.add_separator()
+
+    def messages_from_memory(self, memory):
+        timestamp_key = memory.timestamp
+
+        interaction = self.session.interaction_from_timestamp(timestamp_key)
+        return interaction[0], interaction[1]
+
+    def _last_messages_preview(self, message_history: list[Message], token_limit: int, max_messages: int = 10) -> str:
+        messages = list()
+        token_count = 0
+        sep_tokens = self.count_tokens(text="\n\n")
+        for m in reversed(message_history[-max_messages:]):
+            msg = f"{m.role.upper()}: {truncate(m.content, 500)}"
+            new_token_count = token_count + sep_tokens + self.count_tokens(text=msg)
+            if new_token_count > token_limit:
+                break
+            token_count = new_token_count
+            messages.append(msg)
+        return "\n\n".join(reversed(messages))
+
+    def _update_scratchpad(
+            self, message_history: list[Message], user_message: str, cost_cb: Callable[[float], None],
+            max_changes: int = 10, max_tries: int = 5, max_messages: int = 10, max_scratchpad_tokens: int = 1024,
+    ) -> dict:
+        """An embodied agent interacts with an extremely simple environment to safely update the scratchpad.
+        A rather small scratchpad (1k tokens) seems to work optimally."""
+        assert 0 < max_scratchpad_tokens < self.max_prompt_size // 2
+        ask_kwargs = dict(label="scratchpad", temperature=0, cost_callback=cost_cb, max_messages=max_messages)
+        scratchpad = deepcopy(self.scratchpad)
+        scratchpad_timestamps = self.scratchpad_ts
+        preview_tokens = (self.max_prompt_size // 2) - self.count_tokens(text=sp.to_text(self.scratchpad))
```

Comment:
These two are not used, so remove them at your discretion, depending on whether you're planning to test them afterwards or not.